rfcs/polyfill_testing.md

```diff
@@ -0,0 +1,53 @@
+# RFC 111: Polyfill Testing
+
+## Summary
+
+Add a mechanism for loading one or more polyfill implementations of a web
+platform feature and testing their correctness on existing web platform tests.
+
+## Details
+
+Polyfills are typically best effort implementations of browser features. It's
+not uncommon for them to be outdated, broken or just not to implement all parts
+of the specification they are polyfilling. Additionally, polyfills may only work
+with certain browsers or with certain libraries. This proposal allows testing
+polyfills using the same test suite we test browser compliance with. Being able
+to know which tests the polyfill passes or doesn't pass is useful information
+when developers are determining whether they can rely on it, just as they would
+use this information to know whether they could rely on the feature in the
+browser.
+
+There are a few methods by which we could support this:
+
+1. Rewrite the response from wptserve to insert a script tag which loads the
+   polyfill before running tests.
+2. Add the polyfill into testharnessreport.js.
+3. Add the polyfill into a new /resources/polyfill.js which could also be
+   included by reftests.
+4. Using the [bootstrap scripts
+   feature](https://github.com/w3c/webdriver-bidi/issues/65) once it is
+   available.
+
+## Risks
+
+Depending on the method used there are various risks.
+
+### Rewriting the response
+
+Rewriting the response is a fairly minimal change to the infrastructure, and we
```

Comment:
My concern with this is that injecting the polyfill into all HTML resource loads could lead to tests breaking unexpectedly when they rely on the particular structure of the document e.g. if they assume a certain number of `script` nodes, or that a document has an empty DOM, or similar.

This means that even though there isn't an explicit dependence on the polyfill in the tests, tests do have to be written with a polyfill in mind. I'm not sure this kind of implicit coupling is better than an explicit one, even though it's easier to remove later on.